assignment4/Q1.cpp

```diff
@@ -0,0 +1,126 @@
+#include <iostream>
+#include <vector>
+#include <set>
+
+using namespace std;
+//iterating two times - first to copy the board
+//not sure of its correctness
+//probably something is missing but I do not know yet
```

Comment:
Hi, I made a mistake, sorry to have confused you:
The rank system is only to improve the complexity of the algorithm. It makes sure that you always merge the small set into the large set, preventing a long chain of pointers from one set to another.(it makes sure that the longest chain is O(log(n)) instead of O(n)).

The rank system also prevents circles, but you can use any other method in order to do so: for example, you can choose to always make the lower-index island the parent of the higher-index island. Or you can choose other rules that prevent circles, as the merge rules you chose here.